[ Web ] Register service extensions with DDS, not DWDS

bkonyi opened the pull request:
The ResidentWebRunner was getting an existing VmService instance directly from DWDS, bypassing DDS completely. This change ensures that flutter_tools connects to DDS via websocket instead of using DWDS's internal VmService instance.

Fixes https://github.com/flutter/devtools/issues/2832

bkonyi committed:
[ Web ] Register service extensions with DDS, not DWDS

The ResidentWebRunner was getting an existing VmService instance
directly from DWDS, bypassing DDS completely. This change ensures that
flutter_tools connects to DDS via websocket instead of using DWDS's
internal VmService instance.

Fixes https://github.com/flutter/devtools/issues/2832

bkonyi committed:
Remove whitespace

jonahwilliams reviewed (commented):
Isn't DDS being started by DWDS in this case? Rather than expose a VmService instance that shouldn't be used - should DWDS instead only expose some sort of URI/connector?

It feels like a bit of a sharp edge in this API.

Besides that, we need a way to test this.

bkonyi commented:
> Isn't DDS being started by DWDS in this case? Rather than expose a VmService instance that shouldn't be used - should DWDS instead only expose some sort of URI/connector?
> 
> It feels like a bit of a sharp edge in this API.
> 
> Besides that, we need a way to test this.

If DWDS is going to expose a `VmService` instance, it should expose the correct one (e.g., either the direct connection to DWDS or a web socket connection to DDS). This is more of a starting point for @annagrin to pick things up (I'm heading out for the day shortly) as it was tricky to discover. I figured you'd have some comments too :-)

annagrin commented:
@jonahwilliams debug connection already provides a uri, we can just remove exposing the VMService later in dwds (I need to see how it is used now). Is Ben's change is doing the right thing now regardless, connecting to the exposed uri?

jonahwilliams commented:
Fair enough! I suppose most of this code and dwds was written pre-DDS anyway :)

Commit:
Add tests

Add tests to verify vm functionality on web that devtools rely
in widget explorer.

- test that clients can verify flutter version
- test that clients can call vm.evaluate() without pausing.

annagrin commented on packages/flutter_tools/test/web.shard/vm_service_web_test.dart:
```diff
@@ -0,0 +1,140 @@
+        validateFlutterVersion(client1),
+        validateFlutterVersion(client2)]
+      );
+    }, skip: 'Fails with "Stream already subscribed" exception');
```
This currently fails with "streamListen: (103) Stream already subscribed" exception. Will open a bug and add here.

Commit:
Cleanup

jonahwilliams commented:
You have some conflicts

jonahwilliams reviewed (approved):
LGTM

Commit:
Resolve conflicts

Commit:
Fix analyzer warning

annagrin commented:
@jonahwilliams It looks like the conflict was with your change that added a logger, resolved it. It looks like some code in `devfs_web.dart` still uses the global logger so I left it as is - let me know if you prefer me propagating the logger to `WebDevFS` and `WebAssetServer`.

bkonyi commented on packages/flutter_tools/test/general.shard/resident_web_runner_test.dart:
```diff
@@ -130,7 +130,10 @@ void main() {
+      return ConnectionResult(
+        mockAppConnection,
+        mockDebugConnection,
+        mockDebugConnection.vmService);
```
Nit: at a trailing `,` after `vmService` and put `);` on a newline.

annagrin replied:
done, thanks!

bkonyi reviewed (commented):
Tests LGTM!

Commit:
Adressed CR comments

annagrin reviewed (approved):
Ben's changes LGTM!

jonahwilliams reviewed (approved):
LGTM

DanTup commented:
@bkonyi @annagrin I'm having issues running on Theia today - bisect found this changeset. I'm running Flutter with `--disable-dds` to disable DDS (I was having other issues that I thought may be IPv6-related if I didn't disable it previously, so I'm just disabling it temporarily while working through other issues).

Would this change affect anything when using `--disabled-dds`? The issue I see is that when clicking the Dart Debug extension button, it never seems to connect, and Flutter then prints messages like this:

```
Please click the Dart Debug extension button in the spawned browser window
Connecting to the VM Service is taking longer than expected...
Still attempting to connect to the VM Service...
If you do NOT see the Flutter application running, it might have crashed. The device logs (e.g. from adb or XCode) might have more details.
If you do see the Flutter application running on the device, try re-running with --host-vmservice-port to use a specific port known to be available.
```

DanTup commented:
I've added the verbose output to https://github.com/dart-lang/sdk/issues/45365#issuecomment-815708677, as it seems to be the same error I'd see before when I didn't use `--disable-dds`.
